src/Features/Core/Portable/MetadataAsSource/FileSystemMetadataDocumentPersister.cs

```diff
@@ -0,0 +1,177 @@
+            return true;
+        }
+        catch (Exception ex) when (IOUtilities.IsNormalIOException(ex))
+        {
+            // If we hit an IO exception, we assume the file could not be written.
+            logFailure?.Invoke(ex);
+            return false;
+        }
+    }
+
+    public string GenerateDocumentPath(Guid identifier, string providerName, string fileName)
+    {
+        return Path.Combine(_rootTemporaryPathWithGuid, providerName, identifier.ToString(), fileName);
+    }
+
+    public string ConvertFilePathToDocumentPath(Guid identifier, string providerName, string filePath)
+    {
+        // We already have a file path, and this persister operates on files.  No need to do anything to it.
+        return filePath;
+    }
+
+    public void CleanupGeneratedDocuments()
+    {
+        try
+        {
+            if (Directory.Exists(_rootTemporaryPath))
+            {
+                // Let's look through directories to delete.
+                foreach (var directoryInfo in new DirectoryInfo(_rootTemporaryPath).EnumerateDirectories())
+                {
+                    // Is there a mutex for this one?  If so, that means it's a folder open in another VS instance.
+                    // We should leave it alone.  If not, then it's a folder from a previous VS run.  Delete that
+                    // now.
+                    if (Mutex.TryOpenExisting(CreateMutexName(directoryInfo.Name), out var acquiredMutex))
+                    {
+                        acquiredMutex.Dispose();
+                    }
+                    else
+                    {
+                        TryDeleteFolderWhichContainsReadOnlyFiles(directoryInfo.FullName);
+                    }
+                }
+            }
+        }
+        catch (Exception)
+        {
+        }
+
+        static void TryDeleteFolderWhichContainsReadOnlyFiles(string directoryPath)
+        {
+            try
+            {
+                foreach (var fileInfo in new DirectoryInfo(directoryPath).EnumerateFiles("*", SearchOption.AllDirectories))
+                    IOUtilities.PerformIO(() => fileInfo.IsReadOnly = false);
+
+                IOUtilities.PerformIO(() => Directory.Delete(directoryPath, recursive: true));
+            }
+            catch (Exception)
+            {
+            }
+        }
+    }
+
+    public static SourceText? TryGetExistingTextFromDisk(
+        string documentPath,
+        Encoding encoding,
+        SourceHashAlgorithm checksumAlgorithm,
+        Func<SourceText, bool> verifyExistingDocument)
+    {
+        if (!File.Exists(documentPath))
+            return null;
+
+        using var stream = new FileStream(documentPath, FileMode.Open, FileAccess.Read, FileShare.Read | FileShare.Delete);
+
+        var sourceText = SourceText.From(stream, encoding, checksumAlgorithm, throwIfBinaryDetected: true);
+
+        if (verifyExistingDocument is not null && !verifyExistingDocument(sourceText))
```

Comment:
hunh... i wonder what this is for.

Reply:
this is for sourcelink, to verify the checksum of the document matches the PDB